### PR TITLE
chore: Implement Retries

### PR DESCRIPTION
**Pull request by vicyyn**
# Description

Implement Retries as mentioned in https://github.com/noir-lang/noir/issues/1391
Currently retries maximum of 5 times.

### CODE REVIEW

**Review by jfecher — COMMENTED** [review]
Looks good. My main suggestion is to move out `5` into a constant like `MAX_CRS_DOWNLOAD_RETRIES`

**Review by jfecher — DISMISSED** [review]
👍

File: crates/nargo_cli/src/cli/fs/common_reference_string.rs [hunk 1]
```diff
@@ -35,15 +36,20 @@ pub(crate) fn update_common_reference_string<B: CommonReferenceString>(
...
-    runtime.block_on(fut)
+    for i in 0..MAX_CRS_DOWNLOAD_RETRIES {
+        let fut = if common_reference_string.is_empty() {
+            backend.generate_common_reference_string(circuit)
+        } else {
+            backend.update_common_reference_string(common_reference_string.to_vec(), circuit)
+        };
+
+        let res = runtime.block_on(fut);
+        if res.is_ok() || i == 4 {
```
Comment (phated) [inline crates/nargo_cli/src/cli/fs/common_reference_string.rs:48]:
```suggestion
        if res.is_ok() {
```

File: crates/nargo_cli/src/cli/fs/common_reference_string.rs [hunk 2]
```diff
@@ -35,15 +36,20 @@ pub(crate) fn update_common_reference_string<B: CommonReferenceString>(
...
+            return res;
+        }
+    }
+    unreachable!();
```
Comment (phated) [inline crates/nargo_cli/src/cli/fs/common_reference_string.rs:52]:
```suggestion
   Err(SomeErrorHere)
```

**Review by phated — CHANGES_REQUESTED** [review]
I think we should avoid the unreachable

**Comment by vicyyn** [comment]
I agree @phated! that looks way better 👍

File: crates/nargo_cli/src/cli/fs/common_reference_string.rs (hunk 2 above)
Comment (vicyyn) [reply crates/nargo_cli/src/cli/fs/common_reference_string.rs]: can you elaborate on how would I return the error?
I get this 
```mismatched types
expected associated type `<B as CommonReferenceString>::Error`
            found struct `MyError`
            ``` 
            
every time I try to send an error back. even though my error implements `std::error::Error` similar to `CommonReferenceString`

**Comment by phated** [comment]
@vicyyn did you rebase out your changes on the master branch or something? Ideally you'd be working from separate branches for each PR so you don't have these sync problems.

**Pull request closed without merging**